target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts

```diff
@@ -0,0 +1,330 @@
+		led-upgrade = &led_sys;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,38400";
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		speed {
+			label = "speed";
+			gpios = <&gpio0 19 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_0>;
+		};
+	};
+
+	gpio_fan_sys {
+		compatible = "gpio-fan";
+		alarm-gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
+	};
+
+	gpio_fan_psu_1 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&disable_jtag>;
+		compatible = "gpio-fan";
+
+		alarm-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
+		gpios = <&gpio0 4 GPIO_ACTIVE_LOW>;
+		/* the actual speeds (rpm) are unknown, just use dummy values */
+		gpio-fan,speed-map = <1 0>, <2 1>;
+		#cooling-cells = <2>;
+	};
+
+	gpio_fan_psu_2 {
+		compatible = "gpio-fan";
+		alarm-gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
+	};
+
+	leds {
+		pinctrl-names = "default";
+		compatible = "gpio-leds";
+
+		led_sys: sys {
+			label = "green:sys";
+			gpios = <&gpio0 13 GPIO_ACTIVE_HIGH>;
+		};
+
+		amber_fan {
+			label = "amber:fan";
+			gpios = <&gpio0 16 GPIO_ACTIVE_HIGH>;
+		};
+
+		poe-max {
+			label = "green:poe-max";
+			gpios = <&gpio0 18 GPIO_ACTIVE_HIGH>;
+		};
+
+		speed {
+			label = "green:speed";
+			gpios = <&gpio0 8 GPIO_ACTIVE_HIGH>;
+		};
+
+		poe {
+			label = "green:poe";
+			gpios = <&gpio0 9 GPIO_ACTIVE_HIGH>;
+		};
+
+		green_fan {
+			label = "green:fan";
+			gpios = <&gpio0 15 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	i2c-gpio-0 {
+		compatible = "i2c-gpio";
+		sda-gpios = <&gpio0 2 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		scl-gpios = <&gpio0 1 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		i2c-gpio,delay-us = <2>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+	};
+
+	gpio-restart {
+		compatible = "gpio-restart";
+		gpios = <&gpio0 17 GPIO_ACTIVE_LOW>;
+	};
+
+	virtual_flash {
+		compatible = "mtd-concat";
+
+		devices = <&syspart &usrpart>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x0 0x1000000>;
+			};
+		};
+	};
+};
+
+&gpio0 {
+	poe-enable {
+		gpio-hog;
+		gpios = <23 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "poe-enable";
+	};
+};
+
+&spi0 {
+	status = "okay";
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0xe0000>;
+				read-only;
+			};
+			partition@e0000 {
+				label = "u-boot-env";
+				reg = <0xe0000 0x20000>;
+			};
+			syspart: partition@100000 {
+				label = "sys";
+				reg = <0x100000 0x600000>;
+			};
+			usrpart: partition@700000 {
+				label = "usrimg1";
+				reg = <0x700000 0xa00000>;
+			};
+			partition@1100000 {
+				label = "usrimg2";
+				reg = <0x1100000 0xa00000>;
+			};
```

Comment:
Can you use tplink_sg2xxx_fix_mtdparts() to extend usrimg1 and use the space of both partitions, like we did for 2008P?

https://github.com/openwrt/openwrt/blob/104de8abe455405d43133edc551d50645934d5be/target/linux/realtek/base-files/lib/upgrade/platform.sh#L12-L20

Reply:
I decided against it, because I don't want to overwrite the backup image. A kernel restore should then be sufficient to revert back to stock. Furthermore, I didn't want to change the default baud rate.

Reply:
Then `IMAGE_SIZE` in the build recipe should also be limited to 6M + 10MB, as I see it's currently 26MB.

Reply:
Can users just restore the kernel from a vendor upgrade file, or do they need to take special steps to extract this part? You could add this in the commit message, if you know the steps.

Reply:
This procedure should be verified. I remember trying this on a SG2008P, and the boot process got stuck at "RSA Verify failed!".

Reply:
I didn't try it, no - that's on my ToDo List. I don't think it will work with a vendor upgrade file, but it should work with the backup image I extracted from the IC. Maybe a vendor image could be prepared in a way it is compatible.

Reply:
Since it doesn't make any difference whether usrimg2 is used or not, I extended it to usrimg2 as well. I don't think it's necessary to tell U-Boot about it so I'm not modifying the U-Boot variables. I'll try to come up with a way back to stock - it should be possible to flash kernel + usrimg1 via the TFTP recovery method, at least the one extracted from my flash backup.

Reply:
Once you decrypt the firmware, you can use the following script to separate out the image bits:

https://gist.github.com/mrnuke/2dbcc6afd40c8d5a819457dfff8cdbcf

Reply:
I wrote a similar script literally just now! Mine includes the decryption, though. I'll publish mine once I've tested back-to-stock.

Reply:
Battle of the scripts!

Reply:
Back to stock works fine by simply running `sysupgrade -n -F` with the extracted parts from my script: https://gist.github.com/andyboeh/e5cf74d2912dee527823d654414e4323